tensorboard/data/server/server.rs

```diff
@@ -92,44 +93,122 @@ impl TensorBoardDataProvider for DataProviderHandler {
 
     async fn list_scalars(
         &self,
-        _request: Request<data::ListScalarsRequest>,
+        req: Request<data::ListScalarsRequest>,
     ) -> Result<Response<data::ListScalarsResponse>, Status> {
+        let req = req.into_inner();
+        let want_plugin = parse_plugin_filter(req.plugin_filter)?;
+        let (run_filter, tag_filter) = parse_rtf(req.run_tag_filter);
+        let runs = self.read_runs()?;
+
         let mut res: data::ListScalarsResponse = Default::default();
-        let mut run: data::list_scalars_response::RunEntry = Default::default();
-        run.run_name = "train".to_string();
-        run.tags.push(data::list_scalars_response::TagEntry {
-            tag_name: "accuracy".to_string(),
-            metadata: Some(data::ScalarMetadata {
-                max_step: 5,
-                max_wall_time: FAKE_START_TIME + 5.0,
-                ..Default::default()
-            }),
-            ..Default::default()
-        });
-        res.runs.push(run);
+        for (run, data) in runs.iter() {
+            if !run_filter.want(run) {
+                continue;
+            }
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            let mut run_res: data::list_scalars_response::RunEntry = Default::default();
+            for (tag, ts) in &data.scalars {
+                if !tag_filter.want(tag) {
+                    continue;
+                }
+                let plugin_name = ts
+                    .metadata
+                    .plugin_data
+                    .as_ref()
+                    .map(|pd| pd.plugin_name.as_str());
+                if plugin_name != Some(&want_plugin) {
+                    continue;
+                }
+                let max_step = match ts.valid_values().last() {
+                    None => continue,
+                    Some((step, _, _)) => step,
+                };
+                let max_wall_time = ts
```

Comment:
Maybe worth a TODO to track this on `TimeSeries`?  Would be nice not to have to rescan every single scalar wall time each time we hit `ListScalars`.  I guess it's hard to get this entirely for "free" on `TimeSeries` due to preemptions, but still seems overall cheaper to compute it there (where we could take a cheap path if no preemption happened).